anndata/_core/views.py

```diff
@@ -157,6 +158,68 @@ def as_view_zappy(z, view_args):
     return z
 
 
+try:
+    from ..compat import awkward as ak
+    import weakref
+
+    # Registry to store weak references from AwkwardArrayViews to their parent AnnData container
+    _registry = weakref.WeakValueDictionary()
+    _PARAM_NAME = "_view_args"
+
+    class AwkwardArrayView(_ViewMixin, AwkArray):
+        @property
+        def _view_args(self):
+            """Override _view_args to retrieve the values from awkward arrays parameters.
+
+            Awkward arrays cannot be subclassed like other python objects. Instead subclasses need
+            to be attached as "behavior". These "behaviors" cannot take any additional parameters (as we do
+            for other data types to store `_view_args`). Therefore, we need to store `_view_args` using awkward's
+            parameter mechanism. These parameters need to be json-serializable, which is why we can't store
+            ElementRef directly, but need to replace the reference to the parent AnnDataView container with a weak
+            reference.
+            """
+            parent_key, attrname, keys = self.layout.parameter(_PARAM_NAME)
+            if parent_key is None or attrname is None or keys is None:
+                raise KeyError(
+                    "AwkwardArrayView does not hold reference to original AnnData object."
+                )
+            else:
+                try:
+                    parent = _registry[parent_key]
+                except KeyError:
+                    raise KeyError(
+                        "AwkwardArrayView has invalid reference to original AnnData object."
+                    )
+                else:
+                    return ElementRef(parent, attrname, keys)
+
+        def __copy__(self) -> AwkArray:
+            """
+            Turn the AwkwardArrayView into an actual AwkwardArray with no special behavior.
+
+            Need to override __copy__ instead of `.copy()` as awkward arrays don't implement `.copy()`
+            and are copied using python's standard copy mechanism in `aligned_mapping.py`.
+            """
+            array = self
+            # makes a shallow copy and removes the reference to the original AnnData object
+            array = ak.with_parameter(self, _PARAM_NAME, None)
+            array = ak.with_name(array, None)
+            return array
+
+    @as_view.register(AwkArray)
+    def as_view_awkarray(array, view_args):
+        parent, attrname, keys = view_args
+        parent_key = f"target-{id(parent)}"
+        _registry[parent_key] = parent
+        array = ak.with_parameter(array, _PARAM_NAME, (parent_key, attrname, keys))
+        return ak.with_name(array, name="AwkwardArrayView")
```

Comment:
`ak.with_name` sets a special `__record__` parameter in the array layout. Unlike `ak.with_parameter`, it specifically sets this parameter on the first `ak._v2.contents.RecordArray` that it finds. The `__record__` parameter (and `__array__` parameter) have special meanings in Awkward. Ultimately, they're used to lookup an array-class object from `ak.behavior`, which is the mechanism by which users (and library authors) can customise the interface of an array, e.g. to add custom methods. 

Here, `ak.with_name(array, name="AwkwardArrayView")` is adding the `"AwkwardArrayView"` name to the array, which is associated with the `AwkwardArrayView` class in https://github.com/scverse/anndata/blob/7b571672a300b7db83318fe926d3f8a507946c7d/anndata/_core/views.py#L217

Similarly, the array returned from `ak.with_name(array, None)` effectively has this behavior removed.

Reply:
Yes, Array's can have multiple parameters. You need to overwrite the name, so that you get the correct class wrapper. But other parameters will happily be carried through to arrays modified via `with_name`.

Reply:
I think what @ivirshup meant was: is it somehow possible to stack behaviors? I.e. would it be possible that if a user adds an awkward array to anndata that already has a custom behavior (e.g. the `class Point(ak.Array)` example from the awkward docs) would it be possible that our code makes it an `AwkwardArrayView` that still behaves like a `Point`?

Reply:
Right, that would be where this system gets a little messier. One option would be to create a mixin from the `AwkwardArrayView` and e.g. `Point`, though this would require that the user had registered their behaviour before this code was executed.

As an aside, is it a possibility just to disable mutability, so that Awkward Arrays aren't modified in-place? If so, that could be a feature that we add.

Reply:
tbh, I don't expect this to be very common. We could just raise a `NotImplementedError` if someone tries to make a view of an Array that already has a name attached.

Reply:
I think it would be helpful if I explained the difference between `__array__` and `__record__`, so that we can choose the best approach. By the way, I'm not discounting that we might need to make changes to Awkward Array, but I leave that to @jpivarski as he's currently overseeing the wider vision of the project.

The code that modifies the `Array.__class__` attribute is defined in https://github.com/scikit-hep/awkward/blob/eb19003e7c5e7f227d61bee51066ca1a73dd3302/src/awkward/_v2/_util.py#L479-L491

Simply put:
- `__array__` is looked up on the root layout node, and if found, the class is taken from `ak.behavior[name]`
- `__record__` is looked up recursively through the layout, and if found, the class is taken from `ak.behavior["*", name]`

The `__array__` takes precedence.

Pulling elements out of an array loses the parent parameters, so having `__array__` on the root element has the consequence that `array[0]` would lose the array class. Whereas, if you have a layout node deeper in the tree with `__record__` set, then only items pulled out past that dimension would lose the class. 

`ak.with_name` is special, because it finds a `RecordArray` node to set the `__record__` parameter on. So, `ak.with_name` won't do anything for arrays without `RecordArray` nodes. I think this might need to be an error if this happens, actually, so users don't get confused.

So, in conclusion, I think you want to use `ak.with_parameter(array, `__array__`, name)` instead of `with_name(array, name)`. But, whilst `with_name(array, name)` won't do anything for non-record arrays, `with_parameter(array, "__array__", name)` will break the string abstraction by overwriting `__array__`.

Clearly, breaking strings in a recoverable way is better than just not-working for non-record arrays! Now, how to fix strings:
```python
def get_behavior_name_for(array):
    base = type(array)
    name = f"{base.__name__}View"
    if name in ak.behavior:
        return name
    
    ak.behavior[name] = type(name, (base,), {})
    return name

ak.with_parameter(array, "__array__", get_behavior_name_for(array))
```

At the end of all this, it's clear that Awkward Array is currently friendlier to users adding behaviours than libraries trying to silently modify them. I'll ping @jpivarski again to weigh in on whether there's a better direction to be found here.

Reply:
It should be mentioned that every node in the layout hierarchy can have parameters. As @agoose77 pointed out, `__array__` and `__record__` are special parameters. The `with_name` argument adds a `__record__` parameter to the first record level, a task common enough to need a shortcut. It sounds like your AwkwardArrayView is an array parameter, not a record parameter.

We can illustrate the difference by looking at layouts. Passing `with_name="NamedRecord"` set a parameter on the RecordArray, which is three levels deep here:

```python
>>> ak.Array([[[{"x": 1}]]], with_name="NamedRecord").layout
<ListOffsetArray len='1'>
    <offsets><Index dtype='int64' len='2'>[0 1]</Index></offsets>
    <content><ListOffsetArray len='1'>
        <offsets><Index dtype='int64' len='2'>[0 1]</Index></offsets>
        <content><RecordArray is_tuple='false' len='1'>
            <parameter name='__record__'>'NamedRecord'</parameter>
            <content index='0' field='x'>
                <NumpyArray dtype='int64' len='1'>[1]</NumpyArray>
            </content>
        </RecordArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
```

Assigning a parameter with `ak.with_parameter` puts it on the outermost node:

```python
>>> ak.with_parameter(ak.Array([[[{"x": 1}]]]), "__array__", "NamedArray").layout
<ListOffsetArray len='1'>
    <parameter name='__array__'>'NamedArray'</parameter>
    <offsets><Index dtype='int64' len='2'>[0 1]</Index></offsets>
    <content><ListOffsetArray len='1'>
        <offsets><Index dtype='int64' len='2'>[0 1]</Index></offsets>
        <content><RecordArray is_tuple='false' len='1'>
            <content index='0' field='x'>
                <NumpyArray dtype='int64' len='1'>[1]</NumpyArray>
            </content>
        </RecordArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
```

It's my impression that you want the latter.